test/relayserver/RelayServer.test.ts

```diff
@@ -0,0 +1,1176 @@
+        const userDefinedRelayRequestBody: UserDefinedRelayRequest = {
+          request: {
+            from: owner.address,
+            to: recipient.address,
+            data: encodedData,
+            tokenContract: constants.AddressZero,
+            nonce: 1,
+          },
+          relayData: {
+            callForwarder: constants.AddressZero,
+          },
+        };
+
+        hubInfo.feesReceiver = ethers.Wallet.createRandom().address;
+
+        const envelopingTxRequest = await createEnvelopingTxRequest(
+          userDefinedRelayRequestBody,
+          relayClient,
+          hubInfo
+        );
+
+        expect(() => relayServer.validateInput(envelopingTxRequest)).to.throw(
+          'Wrong fees receiver address'
+        );
+      });
+
+      it('should throw on unacceptable gas price', async function () {
+        const userDefinedRelayRequestBody: UserDefinedRelayRequest = {
+          request: {
+            from: owner.address,
+            to: recipient.address,
+            data: encodedData,
+            tokenContract: constants.AddressZero,
+            nonce: 1,
+          },
+          relayData: {
+            callForwarder: constants.AddressZero,
+          },
+        };
+
+        const envelopingTxRequest = await createEnvelopingTxRequest(
+          userDefinedRelayRequestBody,
+          relayClient,
+          hubInfo
+        );
+
+        envelopingTxRequest.relayRequest.relayData.gasPrice = 0;
+
+        expect(() => relayServer.validateInput(envelopingTxRequest)).to.throw(
+          'Unacceptable gasPrice'
+        );
+      });
+
+      it('should throw on request expired (or too close)', async function () {
+        const userDefinedRelayRequestBody: UserDefinedRelayRequest = {
+          request: {
+            from: owner.address,
+            to: recipient.address,
+            data: encodedData,
+            tokenContract: constants.AddressZero,
+            nonce: 1,
+            validUntilTime: Math.round(Date.now() / 1000),
+          },
+          relayData: {
+            callForwarder: constants.AddressZero,
+          },
+        };
+
+        const envelopingTxRequest = await createEnvelopingTxRequest(
+          userDefinedRelayRequestBody,
+          relayClient,
+          hubInfo
+        );
+
+        expect(() => relayServer.validateInput(envelopingTxRequest)).to.throw(
+          'Request expired (or too close)'
+        );
+      });
+    });
+
+    describe('isTrustedVerifier', function () {
+      it('should not validate if the verifier is trusted', function () {
+        const wrongVerifier = ethers.Wallet.createRandom();
+
+        expect(relayServer.isTrustedVerifier(wrongVerifier.address)).to.be
+          .false;
+      });
+
+      it('should validate if the verifier is trusted', function () {
+        const [verifier] = relayServer.verifierHandler().trustedVerifiers;
+
+        expect(relayServer.isTrustedVerifier(verifier!)).to.be.true;
+      });
+    });
```

Comment:
to be honest, there a some other unit test, but I decided to keep it since there were in the previous file, I just added or refactor tests, but not delete, what do you think @antomor ?

Reply:
As discussed, please let's keep those tests. I don't think they're unit tests since we're not stubbing anything here and furthermore, we're running those tests against regtest (not hardhat). Said that, we'll focus on some e2e tests later on